### PR TITLE
[G&M] React-query for groups components

### PR DESCRIPTION
**Pull request by Guillar1**
image

### CODE REVIEW

File: packages/itwin/grouping-mapping-widget/src/components/Constants.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+
+export enum TErrCodes {
+  QUERY_FETCH_FAILED
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/components/Constants.ts:7]: Where does this come from?

File: packages/itwin/grouping-mapping-widget/src/components/Groups/Editing/GroupAction.tsx [hunk 2]
```diff
@@ -138,34 +128,28 @@ export const GroupAction = (props: GroupActionProps) => {
...
+  const saveGroup = async () => {
+    const accessToken = await getAccessToken();
+    const currentQuery = query || simpleSelectionQuery;
+
+    return props.group
+      ? mappingClient.updateGroup(
+        accessToken,
+        iModelId,
+        props.mappingId,
+        props.group.id ?? "",
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/components/Groups/Editing/GroupAction.tsx:140]: Empty string here will result in a 404.

File: packages/itwin/grouping-mapping-widget/src/components/Groups/Editing/GroupAction.tsx [hunk 3]
```diff
@@ -174,15 +158,43 @@ export const GroupAction = (props: GroupActionProps) => {
...
+  });
 
   const isQueryBuilderStep = currentStep === GroupActionStep.QueryBuilder;
   const isGroupDetailsStep = currentStep === GroupActionStep.GroupDetails;
+
+  const isLoading = isSaveLoading || isQueryLoading;
+
+  const isUpdating = isSaveLoading || isRendering;
+
+  const isBlockingActions = !(
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/components/Groups/Editing/GroupAction.tsx:172]: Couldn't this be just `isRendering || isLoading`?

File: packages/itwin/grouping-mapping-widget/src/components/Groups/GroupsVisualization.tsx [hunk 4]
```diff
@@ -4,18 +4,21 @@
 *--------------------------------------------------------------------------------------------*/
 import type { Group } from "@itwin/insights-client";
 import React, { useCallback, useEffect, useMemo, useState } from "react";
+import type { OverlappedElementGroupPairs } from "../context/GroupHilitedElementsContext";
 import { useGroupHilitedElementsContext } from "../context/GroupHilitedElementsContext";
 import {
-  getHiliteIdsFromGroups,
+  generateOverlappedGroups,
+} from "./groupsHelpers";
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/components/Groups/GroupsVisualization.tsx:11]: Merge the imports. There are 2 separate imports from `./groupsHelpers`.

File: packages/itwin/grouping-mapping-widget/src/components/Groups/GroupsVisualization.tsx [hunk 5]
```diff
@@ -38,152 +46,160 @@ export const GroupsVisualization = ({
...
-    );
-    isNonEmphasizedSelectable && clearEmphasizedElements();
-    handleVisualizationStates(false);
-  }, [handleVisualizationStates, groups, iModelConnection, hiddenGroupsIds, hilitedElementsQueryCache, setNumberOfVisualizedGroups, setOverlappedElementsInfo, setGroupElementsInfo, setOverlappedElementGroupPairs, emphasizeElements, isNonEmphasizedSelectable]);
+    ), [emphasizeElements, hiddenGroupsIds, setNumberOfVisualizedGroups]);
+
+  const zoomToElementsMutation = useMutation({
+    mutationFn: zoomToElements,
+    onSuccess: () => {
+      isNonEmphasizedSelectable && clearEmphasizedElements();
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/components/Groups/GroupsVisualization.tsx:87]: I think this should also use `isMounted()`

File: packages/itwin/grouping-mapping-widget/src/components/Groups/GroupsVisualization.tsx [hunk 6]
```diff
@@ -225,27 +241,30 @@ export const GroupsVisualization = ({
...
-            To get overlap info in detail, click the <Icon><SvgMore/></Icon> button then &ldquo;Overlap Info&rdquo;
+            To get overlap info in detail, click the <Icon><SvgMore /></Icon> button then &ldquo;Overlap Info&rdquo;
           </>
         ) : undefined}
-      </Alert> : undefined,
-  [isAlertClosed, isAlertExpanded, isVisualizing, overlappedElementsInfo.size, showGroupColor]
+      </Alert> : undefined, [isAlertClosed, isAlertExpanded, overlappedElementsMetadata.overlappedElementsInfo.size, showGroupColor]
   );
 
+  const progressConfig = useMemo(() => isVisualizationsEnabled && isResolvingGroupQueries ? { hilitedGroupsProgress: { currentHilitedGroups: groupQueriesProgressCount, totalNumberOfGroups: groups?.length ?? 0 } } : undefined, [groupQueriesProgressCount, groups, isResolvingGroupQueries, isVisualizationsEnabled]);
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/components/Groups/GroupsVisualization.tsx:250]: Maybe make this more readable.

File: packages/itwin/grouping-mapping-widget/src/components/Groups/GroupsVisualizationActions.tsx [hunk 7]
```diff
@@ -27,28 +31,36 @@ export const GroupVisualizationActions = ({
 
   return (
     <ButtonGroup className="gmw-visual-buttons">
-      <ToggleGroupVisibility
-        isLoadingQuery={isLoadingQuery}
-        showGroupColor={showGroupColor}
-        setShowGroupColor={setShowGroupColor}
+      <ToggleSwitch
```
[the remark this answers is not in the document]
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/components/Groups/GroupsVisualizationActions.tsx:34]: I updated the description with the UI change. I mentioned in a status update why I needed to add an additional interaction.

File: packages/itwin/grouping-mapping-widget/src/components/Groups/GroupsVisualizationActions.scss [hunk 8]
```diff
@@ -4,5 +4,6 @@
 *--------------------------------------------------------------------------------------------*/
 
 .gmw-visual-buttons {
+  justify-content: space-between;
```
[the remark this answers is not in the document]
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/components/Groups/GroupsVisualizationActions.scss:7]: It adds the space between the two toggles.

File: packages/itwin/grouping-mapping-widget/src/components/Groups/Editing/GroupAction.tsx (hunk 3 above)
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/components/Groups/Editing/GroupAction.tsx]: De Morgan's Laws applied.

File: packages/itwin/grouping-mapping-widget/src/components/Groups/Editing/GroupAction.tsx (hunk 2 above)
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/components/Groups/Editing/GroupAction.tsx]: Removed

File: packages/itwin/grouping-mapping-widget/src/components/Constants.ts (hunk 1 above)
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/components/Constants.ts]: Added comments for clarification.

File: packages/itwin/grouping-mapping-widget/src/components/Groups/GroupsVisualization.tsx (hunk 6 above)
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/components/Groups/GroupsVisualization.tsx]: Formatted.

File: packages/itwin/grouping-mapping-widget/src/components/Groups/GroupsVisualization.tsx (hunk 4 above)
Comment (Guillar1) [reply packages/itwin/grouping-mapping-widget/src/components/Groups/GroupsVisualization.tsx]: Merged.

**Review by MindaugasButkus — APPROVED** [review]

**Pull request merged**